pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,789 @@
+
+		// If the Calico API server is available, always send the update.
+		// Otherwise, only include an the pool in the update if the pool doesn't exist.
+		if pool, ok := ourPools[p.CIDR]; apiAvailable || !ok || !reflect.DeepEqual(pool.Spec, v1res.Spec) {
+			// The pool either doesn't exist, or it does exist but needs to be updated.
+			if len(currentPools.Items) == 0 {
+				// There are no pools in the cluster. Create them using the v1 API, as they are needed for bootstrapping.
+				// Once the v3 API is available, we'll use that instead. Note that this is an imperfect solution - it still bypasses apiserver validation for
+				// the initial creation of IP pools (although we expect them to be valid due to operator validation). If the bootstrap pools
+				// are invalid and do not enable the Calico apiserver to launch successfully, then manual intervention will be required.
+				toCreateOrUpdate = append(toCreateOrUpdate, v1res)
+			} else if apiAvailable {
+				// The v3 API is available, so use it to create / update the pool.
+				v3res, err := v1ToV3(v1res)
+				if err != nil {
+					r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+					return reconcile.Result{}, err
+				}
+				toCreateOrUpdate = append(toCreateOrUpdate, v3res)
+
+			} else {
+				// The v3 API is not available, and there are existing pools in the cluster. We cannot create new pools until the v3 API is available.
+				// The user may need to manually delete or update pools in order to allow the v3 API to launch successfully.
+				r.status.SetDegraded(operator.ResourceNotReady, "Unable to modify IP pools while Calico API server is unavailable", nil, reqLogger)
+				return reconcile.Result{}, nil
+			}
+		}
+	}
+
+	// Check existing pools owned by this controller that are no longer in
+	// the Installation resource.
+	toDelete := []client.Object{}
+	for cidr, v1res := range ourPools {
+		reqLogger.WithValues("cidr", cidr).V(1).Info("Checking if pool is still valid")
+		found := false
+		for _, p := range installation.Spec.CalicoNetwork.IPPools {
+			if p.CIDR == cidr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// This pool needs to be deleted. We only ever send deletes via the API server,
+			// since deletion requires rather complex logic. If the API server isn't available,
+			// we'll instead just mark the pool as disabled temporarily.
+			reqLogger.WithValues("cidr", cidr, "valid", installation.Spec.CalicoNetwork.IPPools).Info("Pool needs to be deleted")
+			if apiAvailable {
+				// v3 API is available - send a delete request.
+				v3res, err := v1ToV3(&v1res)
+				if err != nil {
+					r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+					return reconcile.Result{}, err
+				}
+				toDelete = append(toDelete, v3res)
+			} else {
+				// v3 API is not available. Just mark the pool as disabled so that new allocations
```

Comment:
This was an attempt at a clever UX improvement, but I am wondering if it's perhaps too complicated and it's better to just error in this case. I can imagine this tripping folks up when they see IP pools in their cluster but they aren't functioning. Maybe better to just say "Can't delete IP pools without the API server" ?

Reply:
I think failing if the API server isn't available could make sense. 

The only reason I can think to not do that is if failing interfered with the API server becoming available. For instance, if we fail here then we wouldn't create / update the IPPools configured in the installation CR and those changes could contain a needed networking fix that would bring the API server back up. Is this a situation that could happen?

We could always deal with attempting to remove IPPools after we've add / updated the ones in the installation CR though.

Reply:
My only concern with disabling a pool if it is removed from the Installation is that (assuming apiserver isn't available/ready) there is different behavior if the user deletes one pool vs deleting one pool and modifying one. In the delete only case it would be disabled as is logical. In the case of deleting and modifying, the deleted one would not be disabled because the controller would return on finding that a pool needed modification.

Reply:
Yeah, it introduces some unwanted intermediate states that you might get stuck in. For example, in my testing I encountered situations where the pool was correctly disabled, but that means you cannot ever get out of that state without manually re-enabling the IP pool because the API server will never launch when there are no enabled IP pools.

I think it's best to just error here.

Reply:
@caseydavenport, do you think it would be a good idea to handle deleting (and subsequently returning an error) for the delete case after the create / modify cases have been handled (since we can successfully bypass the API server if it's down)?

If there was something wrong with the IPPool (I don't know what exactly) that we were trying to delete and it was the cause of the API server being down then could restrict a user from "replacing" the bad IPPool? Maybe there's no way to handle this type of theoretical error without manual intervention?

Reply:
> If there was something wrong with the IPPool (I don't know what exactly) that we were trying to delete and it was the cause of the API server being down then could restrict a user from "replacing" the bad IPPool?

This is desired behavior - if the API server is down we absolutely can't modify or delete IP pools, since we'd be flying blind so to speak.

Reply:
Right, just to double check my understanding we only want to add a slight exception for the start up case where there are no IPPools defined (so allow them to be created), is that correct?

Reply:
Yep, exactly. The only time this controller should do anything with IP pools is:
- If the API server is running.
- There is no API server, and there are no IP pools in the cluster (bootstrapping).